DeepSeek tutorial (#128)
triton-inference-server/tutorials: Popular_Models_Guide/DeepSeek/README.md

# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

In this tutorial we'll use vLLM Backend to deploy
[`DeepSeek-R1-Distill-Llama-8B`](https://huggingface.co/deepseek-ai/DeepSeek-R1-Distill-Llama-8B).
Read more about vLLM [here](https://blog.vllm.ai/2023/06/20/vllm.html) and
the vLLM Backend [here](https://github.com/triton-inference-server/vllm_backend).

## Model Repository

Let's first set up a model repository. In this tutorial we'll use the sample
model repository, provided in the [Triton vLLM backend repository.](https://github.com/triton-inference-server/vllm_backend/tree/main/samples/model_repository/vllm_model)

You can clone the full repository with:
```bash
git clone -b r25.01 https://github.com/triton-inference-server/vllm_backend.git
```

The sample model repository uses [`facebook/opt-125m` model,](https://github.com/triton-inference-server/vllm_backend/blob/80dd0371e0301fabf79c57536e60700d016fcc76/samples/model_repository/vllm_model/1/model.json#L2)
let's replace it with `"deepseek-ai/DeepSeek-R1-Distill-Llama-8B"` and increase

> `gpu_memory_utilization` to `0.9`. Resulted `model.json` should look like:

Reviewer: ```suggestion
`gpu_memory_utilization` to `0.9`. The resulting `model.json` should look like:
```

> `gpu_memory_utilization` to `0.9`. Resulted `model.json` should look like:

Reviewer 2: nit: should we explain why `increase gpu_memory_utilization` is needed and why `0.9`?

Author: clarified

    "disable_log_requests": true,
    "gpu_memory_utilization": 0.9,
    "enforce_eager": true
}
```

## Serving with Triton

Then you can run the tritonserver as usual
```bash
LOCAL_MODEL_REPOSITORY=./vllm_backend/samples/model_repository/
docker run --rm -it --net host --shm-size=2g  --ulimit memlock=-1 \
--ulimit stack=67108864 --gpus all -v $LOCAL_MODEL_REPOSITORY:/opt/tritonserver/model_repository  \
nvcr.io/nvidia/tritonserver:25.01-vllm-python-py3 tritonserver --model-repository=model_repository/
```
The server has launched successfully when you see the following outputs in your console:

```
I0922 23:28:40.351809 1 grpc_server.cc:2451] Started GRPCInferenceService at 0.0.0.0:8001
I0922 23:28:40.352017 1 http_server.cc:3558] Started HTTPService at 0.0.0.0:8000
I0922 23:28:40.395611 1 http_server.cc:187] Started Metrics Service at 0.0.0.0:8002
```

## Sending requests via the `generate` endpoint

As a simple example to make sure the server works, you can use the `generate` endpoint to test. More about the generate endpoint [here](https://github.com/triton-inference-server/server/blob/main/docs/protocol/extension_generate.md).

```bash

> $ curl -X POST localhost:8000/v2/models/vllm_model/generate -d '{"text_input": "What is Triton Inference Server?", "parameters": {"stream": false, n"temperature": 0, "exclude_input_in_output": true, "max_tokens": 45}}' | jq

Reviewer 3: ```suggestion
$ curl -X POST localhost:8000/v2/models/vllm_model/generate -d '{"text_input": "What is Triton Inference Server?", "parameters": {"stream": false, "temperature": 0, "exclude_input_in_output": true, "max_tokens": 45}}' | jq
```

{
  "model_name": "vllm_model",
  "model_version": "1",
  "text_output": " It's a high-performance, scalable, and efficient inference server for AI models. It's designed to handle large numbers of requests quickly and efficiently, making it suitable for real-time applications like autonomous vehicles, smart homes, and more"
}

> ```

Reviewer 3: nit: Could we separate the bash command and the output log? I think this would be easier for users to quickly copy paste the command:

```bash
curl -X POST localhost:8000/v2/models/vllm_model/generate -d '{"text_input": "What is Triton Inference Server?", "parameters": {"stream": false, n"temperature": 0, "exclude_input_in_output": true, "max_tokens": 45}}' | jq
```
Then you should see the output
```text
{
  "model_name": "vllm_model",
  "model_version": "1",
  "text_output": " It's a high-performance, scalable, and efficient inference server for AI models. It's designed to handle large numbers of requests quickly and efficiently, making it suitable for real-time applications like autonomous vehicles, smart homes, and more"
}
```

Author: adjusted

## Sending requests via the Triton client

The Triton vLLM Backend repository has a [samples folder](https://github.com/triton-inference-server/vllm_backend/tree/main/samples)
that has an example client.py to test the model.

```bash

> LOCAL_MODEL_REPOSITORY=./vllm_backend/samples
> docker run -ti --gpus all --network=host --pid=host --ipc=host -v $LOCAL_MODEL_REPOSITORY:/workspace nvcr.io/nvidia/tritonserver:25.01-py3-sdk

Reviewer 4: ```suggestion
LOCAL_WORKSPACE=./vllm_backend/samples
docker run -ti --gpus all --network=host --pid=host --ipc=host -v $LOCAL_WORKSPACE:/workspace nvcr.io/nvidia/tritonserver:25.01-py3-sdk
```

```
Then you can use client as follows:
```bash
python client.py -m vllm_model
```

The following steps should result in a `results.txt` that has the following content

> ```bash

Reviewer: nit: This is just text and not bash:
```suggestion
```
```

Author: adjusted

#  * Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#  * Neither the name of NVIDIA CORPORATION nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

> 

Reviewer 5: # Deploying Deepseek Distill Lllama-8B Models with Triton

Author: added
